### PR TITLE
Issue 4837: Sound is now run on a separate thread

### LINKED ISSUE #4837: A discussion about sound
**Is your feature request related to a problem? Please describe (REQUIRED):**
Defold's sound module never felt usable for me once going past 1 music track and some simple SFX. Here's a list of annoyances off the top of my mind:

  * Last time I checked loading a long music track causes it to load fully into memory, uncompressed PCM. I might be wrong, since this was a long time ago.
  * It's single-threaded and it stutters whenever the main thread is particularly busy.
  * It's impossible to play two sounds in sequence gapless (which would enable the user to create complex conditional loops, which is a must for adaptive music).
  * I can't do a low-pass. Basic DSP support on mixer groups would be very nice. Some dynamic EQ-ing can make things really immersive, for example.

**Describe the solution you'd like (REQUIRED):**
I have the following proposal and I volunteer my spare time towards implementing the following:

Update: Nevermind. I just read through the code and noticed that you're using OpenAL just as a sound output and you generate your own buffers. That's awesome! In that case, switching out OpenAL might not do much in terms of maintainability (might reduce some latency, though). I still stand on my proposal to try my hand at making this threaded.

I'll create separate issues for each of the above annoyances some time tomorrow.

~Let's get rid of OpenAL. I worked with it in the past and it's very limiting. Doing certain things (like streaming) is needlessly complicated, certain things are downright impossible (DSP, for example, unless you use an extension) and implementations are buggy as hell (I remember having issues with the iOS implementation, for example).~

~Instead, I say, looking at Defold's current feature set, it wouldn't be very hard to reimplement it by generating and filling our own sound buffers on a callback from the sound card. It's very easy to port to OS-specific APIs and it allows much more headroom for implementing new features going forward. I propose we use PortAudio on desktop systems and implement it using CoreAudio on iOS and Oboe on Android.~

And I can make it threaded simultaneously, while I'm at it.

**Describe alternatives you've considered (REQUIRED):**
FMOD, but it's not always free and it's a huge overkill library bigger than Defold's executable size by a good factor.

### LINKED ISSUE #4838: Sound threading
**Is your feature request related to a problem? Please describe (REQUIRED):**
Audio playback can become distorted when loading assets or doing high CPU scripts. This requires gamedevs to aggressively mute all audio when for example loading new resources to avoid the user experiencing glitchy sounds. If the user's computer is on the slower end, glitchy audio can still be often heard through normal gameplay, this can often happen on older CPUs or older HDDs since the audio playback is interrupted by the game sim. Making audio play on its own thread would likely solve these.

**Describe the solution you'd like (REQUIRED):**
- Make sound playback on its own thread
Plus via @dapetcu21 these should be considered (and be their own issues)
https://github.com/defold/defold/issues/4837
-  making sure that even after threaded playback implementation, two sounds started in the same frame start simultaneously and sync-up sample-to-sample.
- queueing audio for gapless playback after another sound instance finishes
- being able to query song duration
- being able to query playhead position on a sound instance

https://github.com/defold/defold/blob/869a87718757bdb7bfd6ed078047fa7bfb451de9/engine/sound/README.md

### PR DESCRIPTION
**Pull request by JCash**
* HTML5 sound is still run on the main thread

### CODE REVIEW

File: engine/engine/src/engine.cpp [hunk 1]
```diff
@@ -764,24 +764,16 @@ namespace dmEngine
...
-            ++deviceIndex;
+        sound_params.m_OutputDevice = "default";
+#if defined(__EMSCRIPTEN__)
+        sound_params.m_UseThread = false;
+#else
+        sound_params.m_UseThread = true;
+#endif
+        dmSound::Result soundInit = dmSound::Initialize(engine->m_Config, &sound_params);
+        if (dmSound::RESULT_OK == soundInit) {
+            dmLogInfo("Initialised sound device '%s'\n", sound_params.m_OutputDevice);
```
Comment (JCash) [inline engine/engine/src/engine.cpp:776]: No need to link in extra "backups" which contribute to engine size.

File: engine/engine/src/engine.cpp [hunk 2]
```diff
@@ -764,24 +764,16 @@ namespace dmEngine
...
-            if (dmSound::RESULT_OK == soundInit) {
-                dmLogInfo("Initialised sound device '%s'\n", sound_params.m_OutputDevice);
-                break;
-            }
-            ++deviceIndex;
+        sound_params.m_OutputDevice = "default";
+#if defined(__EMSCRIPTEN__)
+        sound_params.m_UseThread = false;
+#else
+        sound_params.m_UseThread = true;
```
Comment (JCash) [inline engine/engine/src/engine.cpp:772]: By default we want all platforms to use the same behavior.
If needed, we can add a sound config setting later on to control the threaded behavior.

File: engine/gamesys/src/gamesys/scripts/script_sound.cpp [hunk 3]
```diff
@@ -274,19 +268,17 @@ namespace dmGameSystem
      */
     static int Sound_GetGroups(lua_State* L)
     {
-        int top = lua_gettop(L);
+        DM_LUA_STACK_CHECK(L, 1);
+
+        dmhash_t groups[dmSound::MAX_GROUPS];
+        uint32_t count = dmSound::MAX_GROUPS;
+        dmSound::GetGroupHashes(&count, groups);
```
Comment (JCash) [inline engine/gamesys/src/gamesys/scripts/script_sound.cpp:275]: Didn't make sense to have two functions with a lock each.
Replaced with one function, with a total of one lock.

File: engine/sound/src/sound.cpp [hunk 4]
```diff
@@ -194,8 +200,7 @@ namespace dmSound
         dmHashTable<dmhash_t, int> m_GroupMap;
         SoundGroup              m_Groups[MAX_GROUPS];
 
-        Stats                   m_Stats;
-
+        Result                  m_Status;
```
Comment (JCash) [inline engine/sound/src/sound.cpp:203]: Actually mostly used for unit testing.

File: engine/sound/src/sound.cpp [hunk 5]
```diff
@@ -206,8 +211,24 @@ namespace dmSound
         bool                    m_IsDeviceStarted;
         bool                    m_IsPhoneCallActive;
         bool                    m_HasWindowFocus;
+        bool                    m_IsRunning;
     };
 
+    struct OptionalScopedMutexLock
```
Comment (JCash) [inline engine/sound/src/sound.cpp:217]: We want to use the same code with 0 or 1 threads.

**Review by britzl — APPROVED** [review]
Great!

File: engine/sound/src/sound.cpp [hunk 6]
```diff
@@ -1350,6 +1415,23 @@ namespace dmSound
         return RESULT_OK;
     }
 
+    static void SoundThread(SoundSystem* sound)
+    {
+        while (sound->m_IsRunning)
+        {
+            sound->m_Status = UpdateInternal(sound);
+            dmTime::Sleep(8000);
```
Comment (JCash) [inline engine/sound/src/sound.cpp:1423]: For a 60fps game, it'll update ~twice per game frame.
We might want to tweak this value later on, or have a config setting for it.

**Pull request merged**